### DISCUSSION TITLE
Masterpassword length setting

### DISCUSSION THREAD

**Discussion by gamersi** [dani-garcia/vaultwarden#3635] [Ideas]
Hello guys! 
Due to a failed update, I had to reconfigure my whole raspberry pi from scratch. Due to this, I now use Vaultwarden 2023.5.0 instead of 2022.10.something. I noticed an issue: this new version has a minimum password length of 12 characters, but my master password I would like to continue to use hat 10 characters. After looking through the settings and the admin panel, there is no setting for the minimum password length. I would greatly appreciate this setting as the old version had an 8-character minimum afaik. 
Have a lovely rest of your day,
gamersi

**Comment by BlackDex** [comment]
This is build-in into the web-vault, and not a setting we can control on runtime. Only way would be to change it during the web-vault build and use your custom build version. But that would mean you need to do that with every new release of Vaultwarden if also the web-vault was updated.

**Reply by gamersi** [reply]
Ok. I have installed it via the DietPI setup. It uses the standard build. I am too lazy to make my own build, I guess I will have to Sam with it.
Thanks

**Comment by stefan0xC** [comment]
If you downgrade the web vault and set the password you can update to newer versions using your old password but you might not be able to <del>use `argon2id` as KDF (unless you build your own web-vault version)</del> or change your password.
(edit: just tested it and changing it works)

**Comment by xanthos84** [comment]
SERIOUSLY WHO had the glorious idea to implement 12 word passwords as minimum?
I will gladly now stay with official bitwarden as I dont see ANY security advantage of having to type 12 word long passwords to use Vaultwarden.
Thank you Vaultwarden! Really...

**Reply by BlackDex** [reply]
Then you will encounter the same issue, as that is a Bitwarden configured setting ;)

**Reply by zdev-code** [reply]
Guess I'll be using keepass then

**Reply by collse** [reply]
> Then you will encounter the same issue, as that is a Bitwarden configured setting ;)

seems like it is configurable in the enterprise version

**Reply by stefan0xC** [reply]
> seems like it is configurable in the enterprise version

Well, you can require a longer password length via the [master password requirements enterprise policy](https://bitwarden.com/help/policies/#master-password-requirements). But the [`minimumPasswordLength`](https://github.com/bitwarden/clients/blob/19663d958706cb997b2b3dbeeee0c8d18870cf0c/libs/common/src/platform/misc/utils.ts#L37) is hardcoded into the Bitwarden clients to 12.

**Comment by zorgoban** [comment]
This thread is quite old now. I imagine there was enough time to implement this minimum length as a configuration option. What are my options today to keep my password with less than 12 characters?

**Reply by BlackDex** [reply]
There are none. There only way would be to modify the web-vault, and that is not something we will do for the same reasons already mentioned. Also, the other clients are out of our control and would still have this restriction.

**Reply by stefan0xC** [reply]
The 8 character password on my local development instance still works with `web-v2026.2.1`.

**Reply by zorgoban** [reply]
> There only way would be to modify the web-vault, and that is not something we will do for the same reasons already mentioned.

I just read this thread and didn't find any reasons mentioned. Except some hints to clients. But I'm having this issue with the web interface right now. Where exactly is the line of code that controls this? I just tried to create an organization and change the password policy, but even there a minimum length of 12 is enforced. It really seems my switch to vaultwarden failed for now.

> The 8 character password on my local development instance still works with `web-v2026.2.1`.

Good for you.

**Reply by BlackDex** [reply]
First, i find your remark _Good for you_ not very nice and out of scope.

Regarding where it is located.
It's located at several places off I'm correct. Just search the code. It's not our code base, and GitHub search might not find all locations. 

https://github.com/search?q=repo%3Abitwarden%2Fclients+%2Fmin.*%5Cb12%5Cb%2F&type=code

**Reply by zorgoban** [reply]
> First, i find your remark _Good for you_ not very nice and out of scope.

This is a standard reply for "very helpful" comments. It's actually neither friendly nor unfriendly.

**Reply by BlackDex** [reply]
You want a solution that we will not provide, and you do not like that. That is fine of course, and you are allowed to be disappointed. But  your questionn was if 8 char passwords still work, which they do  but you just can't create a new account or update your password with one less than 12.

**Reply by stefan0xC** [reply]
I've already explained how to create an account with a shorter password https://github.com/dani-garcia/vaultwarden/discussions/3635#discussioncomment-6345310 and thus I just corrected your false assumption that an already existing shorter password would not work anymore. You are also free to compile a custom web-vault which has a different `minimumPasswordLength`.

**Reply by zorgoban** [reply]
> I just corrected your false assumption

My only assumptions was about the passage of time.

**Reply by stefan0xC** [reply]
> My only assumptions was about the passage of time.

I don't want to get philosophical but you implied by your question that things have to have changed because of that but neither have we added a configuration option nor do the clients prevent you from keeping a shorter password. While it is possible for organizations to enforce longer passwords on login using policies (which I believe won't affect Admins/Owners) by default the clients won't just enforce the hard-coded minimum for existing accounts. So this minimum is still only enforced when setting up a new password (i.e. when you want to change your password or when you create a new account). And therefore the already discussed options to create an account with a shorter password are still viable (even if we don't recommend them).

**Comment by spmfox** [comment]
Hello - I'd like to add to this discussion. A few questions: Is the reason for this change to match Bitwarden client requirements? Is there an indication that the clients will block users from /using/ a shorter password?

While I completely understand the nature of not wanting to deviate from Bitwarden, it could make things confusing and messy, I also think that little changes like this are what bring out the value of Vaultwarden.

I do not see the reason why the web client couldn't allow slight changes that differ, if there is  a user need, as long as the Bitwarden clients do not have a problem with this. Even if the setting is buried maybe in the admin panel or environment variable etc.

For me personally, as a rule I do not like shorter passwords however I have a very particular use case and I'm using a token device for 2FA. I would like the ability to take off the guard rails and use the shorter password because I understand the risks in my scenario.

Thank you for your time and effort with this amazing project.

**Reply by BlackDex** [reply]
The web-vault is from Bitwarden. We only make minor changes like logos and names. The rest stays as is.

These settings are built-in into the clients, that also means the mobile, and desktop clients which use the same hardcoded defaults.

**Comment by M1CK431** [comment]
Not a "bulletproof" solution, but it works for now and doesn’t require rebuilding anything or downgrading to an older version of the web app:
1. Locate the web app directory (in my case, it’s at `/usr/share/webapps/vaultwarden-web`).
2. `cd /usr/share/webapps/vaultwarden-web/app`
3. Run: `sed -i 's/minimumPasswordLength=12/minimumPasswordLength=8/g' main.*.js`
4. Reload the web app in your browser, log in, and change your password 🎉

Note that since this is a **client-side** limitation, a more proper approach would be to use the backend API directly (e.g., via `curl` or [vaultwarden-cli](https://github.com/haydonryan/vaultwarden-cli)). I didn’t test that method myself, as I prefer to allow shorter passwords for all users on my Vaultwarden instance anyway.

Since this modifies frontend files, **Vaultwarden updates will overwrite `main.*.js`**. If you want it to persist, you could:
- Add the `sed` command to a post-update hook or systemd service
- Use a reverse proxy to serve a custom JS file
- Or simply document that it needs to be reapplied after each update

**Comment by brlin-tw** [comment]
Hello all, also encountered this issue and I ended up workaround it using the following NGINX configuration:

```nginx
    location ~ ^/app/main\.[a-f0-9]+\.js$ {
        proxy_pass http://127.0.0.1:8080;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto https;
        proxy_http_version 1.1;

        # Workaround minimun password length requirement too strict problem
        sub_filter 'minimumPasswordLength=12' 'minimumPasswordLength=8';
        sub_filter_once off;
        sub_filter_types text/javascript;
    }
```

this configuration should survive service upgrade as long as the related code doesn't change too much, cheers!

**Reply by LukasHen111** [reply]
That worked for me!
